runtime/src/main/java/org/corfudb/util/retry/MixedBoundRetry.java

```diff
@@ -0,0 +1,132 @@
+ * backoff strategy is used, which subsequently updates the lastDuration. The parameter overallMaxRetryDuration
+ * is set to limit the overall retry duration. After the overallMaxRetryDuration, the RetryExhaustedException
+ * will be thrown inside of the run() method.
+ */
+public class MixedBoundRetry<E extends Exception, F extends Exception,
+        G extends Exception, H extends Exception, O> extends AbstractRetry<E, F,
+        G, H, O, MixedBoundRetry> {
+
+    /**
+     * Overall duration of the retry
+     */
+    @Setter
+    @Getter
+    @NonNull
+    Duration overallMaxRetryDuration = Duration.ofSeconds(5);
+
+    /**
+     * The max duration of a single retry, after which the retry duration will stop growing.
+     */
+    @Setter
+    @Getter
+    @NonNull
+    Duration maxRetryDuration = Duration.ofSeconds(2);
+
+    /**
+     * The initial duration for the interval retry and the base duration for the exponential backoff retry.
+     */
+    @Setter
+    @Getter
+    @NonNull
+    Duration baseDuration = Duration.ofSeconds(1);
+
+    /**
+     * The base of the exponential backoff growth function
+     */
+    @Setter
+    @Getter
+    int multiplier = 2;
+
+    /**
+     * Randomization factor from 0-1 to adjust the exponential backoff duration.
+     */
+    @Setter
+    @Getter
+    float randomPart = 0f;
+
+    /**
+     * Last waited duration
+     */
+    @Getter
+    @NonNull
+    Duration lastDuration = baseDuration;
+
+    /**
+     * Number of backoff retries requested
+     */
+    @Getter
+    int backOffCounter = 0;
+
+    /**
+     * When will the retry loop end
+     */
+    @Getter
+    long endSession = 0;
+
+
+    public MixedBoundRetry(IRetryable<E, F, G, H, O> runFunction) {
+        super(runFunction);
+    }
+
+
+    @Override
+    public void nextWait() {
+        if (System.currentTimeMillis() + lastDuration.toMillis() > endSession) {
+            throw new RetryExhaustedException();
+        }
+        if (maxRetryDuration.toMillis() > 0) {
+            lastDuration = Duration.ofMillis(Math.min(lastDuration.toMillis(), maxRetryDuration.toMillis()));
+        }
+        Sleep.sleepUninterruptibly(lastDuration);
+    }
+
+    private void backOffWait() {
+        backOffCounter += 1;
+        long sleepTime = baseDuration.toMillis() * (long) Math.pow(multiplier, backOffCounter);
+        sleepTime -= sleepTime * new Random().nextFloat() * randomPart;
+        if (maxRetryDuration.toMillis() > 0) {
+            sleepTime = Math.min(sleepTime, maxRetryDuration.toMillis());
+        }
+        if (System.currentTimeMillis() + sleepTime > endSession) {
+            throw new RetryExhaustedException();
+        }
+        lastDuration = Duration.ofMillis(sleepTime);
+        Sleep.sleepUninterruptibly(lastDuration);
+    }
+
+    @Override
+    public O run() throws E, F, G, H, InterruptedException {
+        endSession = System.currentTimeMillis() + overallMaxRetryDuration.toMillis();
```

Comment:
This clock is not monotonic, what is the failure mode of this code if the timestamp regresses ?

Consider using a scheduled executor since it accounts for these edge timing issues.

Reply:
Fixed